### Parte 3 – Cruce por ubigeo y análisis

#### Paso 1
Crear el archivo datos/ubigeos.csv con las columnas ubigeo y departamento

In [12]:
import pandas as pd
import os

ubigeos = pd.DataFrame({
    "ubigeo": ["01", "02", "03", "04", "05", "06", "07", "08", "09", "10",
               "11", "12", "13", "14", "15", "16", "17", "18", "19", "20",
               "21", "22", "23", "24", "25"],
    "departamento": ["Amazonas", "Áncash", "Apurímac", "Arequipa", "Ayacucho",
                     "Cajamarca", "Callao", "Cusco", "Huancavelica", "Huánuco",
                     "Ica", "Junín", "La Libertad", "Lambayeque", "Lima",
                     "Loreto", "Madre de Dios", "Moquegua", "Pasco", "Piura",
                     "Puno", "San Martín", "Tacna", "Tumbes", "Ucayali"]
})

os.makedirs("datos", exist_ok=True)

ubigeos.to_csv("datos/ubigeos.csv", index=False, encoding="utf-8-sig")

ubigeos

,ubigeo,departamento
0,01,Amazonas
1,02,Áncash
2,03,Apurímac
3,04,Arequipa
4,05,Ayacucho
5,06,Cajamarca
6,07,Callao
7,08,Cusco
8,09,Huancavelica
9,10,Huánuco


#### Paso 2
Leer archivo normal y otra con *dtype={"ubigeo": str}*. Mostrar la diferencia y explicar la importancia

In [13]:
# Lectura normal
ubigeo_0= pd.read_csv("datos/ubigeos.csv")
print("Lectura normal:")
print(ubigeo_0.head())
print()
print(ubigeo_0.dtypes)

# Lectura con dtype
ubigeo = pd.read_csv("datos/ubigeos.csv", dtype={"ubigeo": str})
print("\nLectura con dtype:")
print(ubigeo.head())
print()
print(ubigeo.dtypes)

Lectura normal:
   ubigeo departamento
0       1     Amazonas
1       2       Áncash
2       3     Apurímac
3       4     Arequipa
4       5     Ayacucho

ubigeo          int64
departamento      str
dtype: object

Lectura con dtype:
  ubigeo departamento
0     01     Amazonas
1     02       Áncash
2     03     Apurímac
3     04     Arequipa
4     05     Ayacucho

ubigeo          str
departamento    str
dtype: object


Ambas lecturas muestran una diferencia en la variable ubigeo. En la lectura normal, se importa como número entero y se borra el 0 que se había colocado al inicio en los casos que tenían 1 dígito. Mientras que, en la lectura con dtype={"ubigeo": str}, como se establece en específico que se considere a la variable "ubigeo" como string, se mantiene la variable como string y no se modifica. 

**Importancia:** Identificar esta diferencia es importante ya que permite visualizar la relevancia de especificar el tipo de variable al momento de importarlo, en especial en casos donde la variable podría calificar como diferentes tipos y no esté muy claro cómo se quisiera que se considere, como el caso de la variable "ubigeo". Además, si luego se quiere unir una base con otra a través de esa variable es importante que estén en el mismo formato para que no se presenten inconvenientes.

### Paso 3
Agregar ubigeo a decretos_por_departamento.csv y a lluvias_por_departamento.csv

In [14]:
# Bases originales
decretos= pd.read_csv("datos/decretos_por_departamento.csv")
lluvias= pd.read_csv("datos/lluvias_por_departamento.csv")

display(decretos.head())
display(lluvias.head())

,departamento,declaratorias,prorrogas
0,Amazonas,4,2
1,Áncash,5,5
2,Apurímac,7,4
3,Arequipa,7,6
4,Ayacucho,7,5


,departamento,capital,latitud,longitud,lluvia_total_mm,dias_lluvia_fuerte
0,Loreto,Iquitos,-3.74814,-73.25290,1361.8,19
1,Junín,Huancayo,-12.06866,-75.21027,1276.9,11
2,Madre de Dios,Puerto Maldonado,-12.58946,-69.19948,1149.5,15
3,Áncash,Huaraz,-9.52614,-77.52869,1058.3,3
4,Ucayali,Pucallpa,-8.37915,-74.55387,887.4,9


In [15]:
# Agregar ubigeo por nombre de departamento
decretos_ubi = decretos.merge(ubigeo, on="departamento", how="left")
lluvias_ubi = lluvias.merge(ubigeo, on="departamento", how="left")

display(decretos_ubi.head())
display(lluvias_ubi.head())

,departamento,declaratorias,prorrogas,ubigeo
0,Amazonas,4,2,01
1,Áncash,5,5,02
2,Apurímac,7,4,03
3,Arequipa,7,6,04
4,Ayacucho,7,5,05


,departamento,capital,latitud,longitud,lluvia_total_mm,dias_lluvia_fuerte,ubigeo
0,Loreto,Iquitos,-3.74814,-73.25290,1361.8,19,16
1,Junín,Huancayo,-12.06866,-75.21027,1276.9,11,12
2,Madre de Dios,Puerto Maldonado,-12.58946,-69.19948,1149.5,15,17
3,Áncash,Huaraz,-9.52614,-77.52869,1058.3,3,02
4,Ucayali,Pucallpa,-8.37915,-74.55387,887.4,9,25


### Paso 4
Mostrar los departamentos que no encontraron su ubigeo y corregir hasta que no quede ninguno 

In [16]:
# Archivo decretos por departamento
print("Filas",len(decretos),"->",len(decretos_ubi))
print("No cruzó: ",decretos_ubi["ubigeo"].isna().sum())
print(decretos_ubi[decretos_ubi["ubigeo"].isna()].head()) #vacío, todos cruzaron

Filas 25 -> 25
No cruzó:  0
Empty DataFrame
Columns: [departamento, declaratorias, prorrogas, ubigeo]
Index: []


In [17]:
# Archivo lluvias por departamento
print("Filas",len(lluvias),"->",len(lluvias_ubi))
print("No cruzó: ",lluvias_ubi["ubigeo"].isna().sum())
print(lluvias_ubi[lluvias_ubi["ubigeo"].isna()].head()) #vacío, todos cruzaron

Filas 25 -> 25
No cruzó:  0
Empty DataFrame
Columns: [departamento, capital, latitud, longitud, lluvia_total_mm, dias_lluvia_fuerte, ubigeo]
Index: []


En ambas bases lograron cruzarse todas las filas con la base ubigeo

### Paso 5
Cruzar las dos tablas por ubigeo

In [18]:
# Quitar el nombre en decretos para no duplicar la columna
decretos_cruce = decretos_ubi.drop(columns="departamento")

# Cruce por ubigeo de las dos tablas
tabla_final = lluvias_ubi.merge(
    decretos_cruce, on="ubigeo", how="outer", indicator=True
)

tabla_final.head()

,departamento,capital,latitud,longitud,lluvia_total_mm,dias_lluvia_fuerte,ubigeo,declaratorias,prorrogas,_merge
0,Amazonas,Chachapoyas,-6.23169,-77.86903,455.0,0,01,4,2,both
1,Áncash,Huaraz,-9.52614,-77.52869,1058.3,3,02,5,5,both
2,Apurímac,Abancay,-13.63426,-72.88380,670.6,2,03,7,4,both
3,Arequipa,Arequipa,-16.39899,-71.53747,235.3,1,04,7,6,both
4,Ayacucho,Ayacucho,-13.16380,-74.22345,467.5,1,05,7,5,both


### Paso 6
Verificar que la tabla final tenga 25 filas y que departamentos sin declaratoria aparezcan con 0, no desaparezcan

In [19]:
# Verificación del N° de filas y origen de las filas
print("Filas:", len(tabla_final))   # debe ser 25
print("\nOrigen de las filas:")
print(tabla_final["_merge"].value_counts())

Filas: 25

Origen de las filas:
_merge
both          25
left_only      0
right_only     0
Name: count, dtype: int64


Se tienen 25 filas y todas cruzaron

In [22]:
# Verificación de los valores de las columnas de declaratorias y prorrogas para identificar si hay valores faltantes
def tabla_frecuencia(serie):
    tabla = serie.value_counts(dropna=False).sort_index().to_frame("frecuencia")
    tabla.index = tabla.index.map(lambda x: "Sin dato (NA)" if pd.isna(x) else int(x))
    tabla.index.name = "valor"
    tabla.loc["Total"] = tabla["frecuencia"].sum()
    return tabla

print("Frecuencia de declaratorias:")
display(tabla_frecuencia(tabla_final["declaratorias"]))

print("\nFrecuencia de prorrogas:")
display(tabla_frecuencia(tabla_final["prorrogas"]))

Frecuencia de declaratorias:


,frecuencia
valor,
0,4
2,1
3,6
4,3
5,4
6,2
7,5
Total,25



Frecuencia de prorrogas:


,frecuencia
valor,
0,4
1,1
2,7
3,5
4,3
5,4
6,1
Total,25


No hay valores faltantes, si es que no se tienen declaratorias están como valor 0

### Paso 7
Guardar base de datos, asegurar previamente que la variable ubigeo esté como texto de 2 dígitos

In [23]:
# Borrar la columna "_merge" que indica el origen de las filas
tabla_final = tabla_final.drop(columns="_merge")
print(tabla_final.columns)

Index(['departamento', 'capital', 'latitud', 'longitud', 'lluvia_total_mm',
       'dias_lluvia_fuerte', 'ubigeo', 'declaratorias', 'prorrogas'],
      dtype='str')


In [24]:
# Asegurar que ubigeo sea texto de 2 dígitos
tabla_final["ubigeo"] = tabla_final["ubigeo"].astype(str).str.zfill(2)

# Colocar primero la columna ubigeo y luego las demás columnas
columnas = ["ubigeo"] + [col for col in tabla_final.columns if col != "ubigeo"]
tabla_final = tabla_final[columnas]

# Ordenar por ubigeo
tabla_final = tabla_final.sort_values("ubigeo").reset_index(drop=True)

# Guardar la tabla final en un archivo CSV
tabla_final.to_csv("datos/tabla_final.csv", index=False, encoding="utf-8-sig")

display(tabla_final.head())

,ubigeo,departamento,capital,latitud,longitud,lluvia_total_mm,dias_lluvia_fuerte,declaratorias,prorrogas
0,01,Amazonas,Chachapoyas,-6.23169,-77.86903,455.0,0,4,2
1,02,Áncash,Huaraz,-9.52614,-77.52869,1058.3,3,5,5
2,03,Apurímac,Abancay,-13.63426,-72.88380,670.6,2,7,4
3,04,Arequipa,Arequipa,-16.39899,-71.53747,235.3,1,7,6
4,05,Ayacucho,Ayacucho,-13.16380,-74.22345,467.5,1,7,5


### Paso 8
Realizar 2 gráficos

In [25]:
import plotly.express as px
periodo = "Periodo: 1 de enero al 31 de mayo de 2024"

In [26]:
# Gráfico de dispersión
grafico_1 = px.scatter(
    tabla_final,
    x="lluvia_total_mm",
    y="declaratorias",
    hover_name="departamento",
    title="Lluvia total (mm) vs declaratorias de emergencia, por departamento"
            f"<br><sup>{periodo}</sup>",
    labels={"lluvia_total_mm": "Lluvia total (mm)",
            "declaratorias": "Número de declaratorias"}
)

grafico_1.show()

In [27]:
# Gráfico de barras
top10 = tabla_final.sort_values("declaratorias", ascending=False).head(10)

grafico_2 = px.bar(top10, x="departamento", y="declaratorias",
                   title="Top 10 departamentos con más declaratorias de emergencia"
                            f"<br><sup>{periodo}</sup>",
                    labels={"departamento": "Departamento", "declaratorias": "Número de declaratorias"})
grafico_2.show()

In [28]:
# Validar si se presentan empates a final del top 10 con los que se quedan fuera
ranking_decretos = tabla_final.sort_values("declaratorias", ascending=False)[["departamento", "declaratorias"]]
ranking_decretos.insert(0, "puesto", range(1, len(ranking_decretos) + 1))
ranking_decretos.set_index("puesto")

,departamento,declaratorias
puesto,,
1,Apurímac,7
2,Cusco,7
3,Arequipa,7
4,Ayacucho,7
5,Huancavelica,7
6,Lima,6
7,Huánuco,6
8,Áncash,5
9,Junín,5


Se presenta un empate entre el 10 y el 11 de 5 declaratorias, en ese caso sería mejor mostrar el gráfico de barras hasta el 11

In [29]:
# Gráfico de barras ajustado por empate al final del top 10
top11 = tabla_final.sort_values("declaratorias", ascending=False).head(11)

grafico_2v2 = px.bar(top11, x="departamento", y="declaratorias",
                   title="Top 11 departamentos con más declaratorias de emergencia"
                            f"<br><sup>{periodo}</sup>",
                    labels={"departamento": "Departamento", "declaratorias": "Número de declaratorias"})
grafico_2v2.show()

### Paso 9
Responder preguntas de análisis, limitaciones y conclusión

In [32]:
# Datos adicionales para responder preguntas

# i
#---#
print("Los 5 departamentos con mayor cantidad de declaratorias:")
display(ranking_decretos.set_index("puesto").head(5))

ranking_lluvias = tabla_final.sort_values("lluvia_total_mm", ascending=False)[["departamento", "lluvia_total_mm"]]
ranking_lluvias.insert(0, "puesto", range(1, len(ranking_lluvias) + 1))
print("Los 3 departamentos con más lluvia total (mm):")
display(ranking_lluvias.set_index("puesto").head(3))

# ii
#---#
print("Los 3 departamentos con menos lluvia total (mm):")
display(ranking_lluvias.set_index("puesto").tail(3))

# iii
#---#
decretos_tipo = pd.read_csv("datos/decretos_lluvias.csv")

pd.crosstab(decretos_tipo["motivo"], decretos_tipo["tipo"],
            margins=True, margins_name="Total")


Los 5 departamentos con mayor cantidad de declaratorias:


,departamento,declaratorias
puesto,,
1,Apurímac,7
2,Cusco,7
3,Arequipa,7
4,Ayacucho,7
5,Huancavelica,7


Los 3 departamentos con más lluvia total (mm):


,departamento,lluvia_total_mm
puesto,,
1,Loreto,1361.8
2,Junín,1276.9
3,Madre de Dios,1149.5


Los 3 departamentos con menos lluvia total (mm):


,departamento,lluvia_total_mm
puesto,,
23,Ica,19.3
24,Callao,17.3
25,Lima,17.3


tipo,declara,prorroga,Total
motivo,,,
impacto de daños,10,8,18
peligro inminente,0,2,2
Total,10,10,20


**i. ¿Cuáles fueron los 3 departamentos con más declaratorias? ¿Coinciden con los 3 donde más llovió?**

Se presentó un empate entre los primeros puestos en el ranking por número de declaratorias, 5 departamentos presentaron la mayor cantidad que fue 7: Apurímac, Arequipa, Ayacucho, Cusco y Huancavelica. No coinciden con los 3 departamentos donde más llovió: Loreto, Junín y Madre de Dios.

**ii. ¿Hay departamentos con mucha lluvia y pocas o ninguna declaratoria, o al revés? Den al menos dos razones posibles.**

Como se puede observar en el gráfico de dispersión realizado en el paso 8, se presentan departamentos con mucha lluvia y pocas declaratorias, y otros con varias declaratorias y poca lluvia. Por ejemplo, Loreto registró la mayor lluvia total (1361.8 mm), pero solo tuvo 3 declaratorias, menos que departamentos con mucha menos lluvia. En el otro extremo, Lima tuvo 6 declaratorias, a pesar de ser uno de los departamentos con menor lluvia total (17.3 mm).

Algunas razones posibles son:

- **Medición de la lluvia solo en la capital.** En Lima, por ejemplo, se calculó de la ciudad de Lima, en la costa, donde casi no llueve; pero en las provincias de sierra del departamento si se podrían haber presentado lluvias intensas que generaron daños.
- **Vulnerabilidad y lluvia anómala.** En Loreto las lluvias intensas son habituales y el territorio podría estar más adaptado a ellas; en cambio, lluvias moderadas en zonas secas o con quebradas podrían haber causado daños graves.
- **Ámbito territorial de los decretos.** Un mismo departamento podría haber recibido varias declaratorias para distintos grupos de distritos, lo que podría elevar el conteo aunque la lluvia no necesariamente haya sido mayor.
- **Declaratorias preventivas.** Las declaratorias por peligro inminente se emiten según pronósticos, antes del evento, por lo que no dependen de la lluvia que finalmente cayó.

**iii. ¿Qué porcentaje de las declaratorias fue por peligro inminente (antes del daño) y qué porcentaje por impacto de daños (después)? ¿Qué dice eso sobre la gestión del riesgo en su temporada?**

El 100% de las declaratorias se emitió por impacto de daños (después del daño) y ninguna por peligro inminente (antes del daño). En las prórrogas, el 80% fue por impacto de daños y el 20 % por peligro inminente. En conjunto, el 90% responde a daños ya ocurridos y solo el 10% a peligro inminente.

Esto sugiere que, en la temporada analizada, la gestión del riesgo mediante estados de emergencia fue principalmente reactiva: el Estado declaró emergencia una vez ocurridos los daños, más que para anticiparse a ellos.

**iv. Mencionen al menos tres limitaciones de este análisis.**

- **La lluvia se midió solo en la capital de cada departamento.** Un departamento abarca zonas que pueden presentar diferentes climas; por ejemplo, la ciudad de Lima casi no registra lluvia, pero en las provincias de sierra del departamento podría presentarse en mayor magnitud.
- **Medición de lluvia en mm total y no consideración de lluvias anómalas** Departamentos en la selva, como Loreto, presentan mucha lluvia de forma más constante y podrían estar más preparados para ello que en zonas donde no se genera de forma muy constante.
- **El análisis se realizó por departamento, no a nivel distrital**. Si se desagrega a nivel distrital, se podría realizar una mejor evaluación, ya que permite diferenciar las zonas más afectadas.

**v. Conclusión: respondan la pregunta del trabajo en 5 a 8 líneas.**

*Pregunta del trabajo: ¿El Estado declara emergencia donde más llueve?*

En la temporada de lluvias de 2024 (1 de enero al 31 de mayo), el Estado no necesariamente declaró emergencia donde más llovió. Por ejemplo, Loreto fue el departamento con más lluvia y tuvo solo 3 declaratorias, mientras que Lima tuvo 6 con muy poca lluvia. Además, todas las declaratorias se emitieron por impacto de daños, lo que indica que el Estado actúa principalmente después de que ocurren los daños y no tanto de manera preventiva. Esto sugiere que las declaratorias responden más a la vulnerabilidad del territorio y a los daños ocurridos que a las proyecciones sobre lluvias. Sin embargo, por las limitaciones señaladas, como medir la lluvia solo en la capital de cada departamento, no considerar lluvias anómalas y evaluar a nivel departamental y no distrital, se requiere incorporar estas variables para realizar una mejor evaluación.